app/commands/mentor/discussion/process_finished.rb

```diff
@@ -4,14 +4,16 @@ class Mentor::Discussion::ProcessFinished
   initialize_with :discussion
 
   def call
```

Comment:
There were a lot of potential race conditions in here. I've moved the logic of badges and roles into the `UpdateStats` call, as we can more firmly specify the order there.